runtime/test/slot-composer-tests.js

```diff
@@ -209,4 +210,107 @@ recipe
     assert.deepEqual({'id1': ['C moved to id1'], 'id2': ['B moved to id2']}, gatherRenderings(itemSlotContext));
   });
 
+  it('renders inner slots in transformations without intercepting', async () => {
+    const {arc, slotComposer} = await TestHelper.create({
+      manifestString: `
+        particle TransformationParticle in 'TransformationParticle.js'
+          consume root
+    
+        recipe
+          slot 'rootslotid-root' as slot0
+          TransformationParticle
```

Comment:
what if i want the TransformationParticle to `provide set of detail` for other transformation particles to consume? 
so that we can also do
```
recipe
  slot 'rootslotid-root' as slot0
  TransformationParticle
     hostedParticle = A
     consume root as slot0
         // provide set of content ??
         // provide set of detail ??
  TransformationParticle
     hostedParticle = B
     consume set of detail // as ??
```

Specifically the `content` slot is created by TransformationParticle, but it is not present in the manifest, what if another hosted particle wants to consume it?
This TransformationParticle example works well because it is specially tailored for particles A and B, but I'm not sure how it can be generalized.

Reply:
If the TransformationParticle wants to `provide set of detail` then it needs to list that fact and do the internal plumbing (createSlot calls) necessary for it to happen. This then gets hooked up to the internal particles by standard recipe wrangling.

Am I missing some detail as to why this wouldn't work, though?

Reply:
I'm not saying it won't, I'm mostly curious about how `content` slot should be treated in manifest - should it be provided? and consumed? Seems to me that the TransformationParticle (with hostedParticle = A) both - provides and consumes it, but it would look weird in a manifest.
But if we don't declare it, we can't restrict the handles that are rendered in the slot.

Reply:
Seems to me that:
`content` slot is a hosted slot. That's between TransformationParticle and A, i.e. A renders to it, and TransformationParticle intercepts these renders calls and can do with them as it pleases. No other particle can see it, unless TransformationParticle instantiates another particle that renders to it.

`detail` slot is interesting. I am not sure whether what Maria proposes to do would work right now - my intuition is it won't work, because the detail slot provided by A would be a different slot context than a detail slot provided by TransformationParticle - i.e. we would have 2 separate objects representing these slots.

Shane was mentioning his ideas for mapping slots. I am thinking whether this is applicable here? Particle A is living inside a transformation. Should the transformation explicitly map A.detail to TransformationParticle.detail to expose this slot outside a transformation?

Reply:
I think what Maria is talking about will not work yet.

I think exposing a slot provided by a particle inside a transformation to the world outside a transformation should be explicitly done (via something like `mapSlot`?) and not relying on names alone. It should be possible for particle A (in a transformation) to provide `detail`, which is consumed by B (also in a transformation), while at the same time transformationParticle could provide a `detail` slot, where the name clash is entirely coincidental, and transformationParticle.detail !== A.detail. This requires some more thinking and refactoring, which I pointed out in a TODO in code. The code in this patch relies on non-namespaced names, which will only work for simple use cases.

Let me file a bug for myself and take a stab at this at a later stage. I hope that sounds good.

Reply:
Anything that mapSlot will be able to do should be possible with createSlot as well. I don't understand why we can't have the provide affordance listed (i.e. there's a real provide slot on the transformation)? Can't the transformation particle create individual slots for each inner particle, then ensure that a matching set slot tag appears in its rendering?

Reply:
I've checked and it works, like Shane describes.

I've enhanced the test case (code not included here) with a particle C that lives outside the transformation and consume a `detail` slot which TransformationParticle now declares to provide. Both B and C will render into the `detail` slot originally in the template produced by `A` and passed through to the rendering of a `TransformationParticle`.
 
However, it probably doesn't do what we would like it to do. We end up with 2 SlotContexts using the same DOM element as a container (container keyed by the slotName _detail_):
- One `HostedSlotContext` created by HostedSlotConsumer corresponding to `A`. Particle `B` also inside a transformation writes to this.
- One regular `SlotContext`, created by the SlotDomConsumer corresponding to the transformation particle. Particle `C` writes to this.

It works initially, but will cause issues when for example one of contexts clears the container.

What do you Shane think of unintentional name clashes that I described here:

> I think exposing a slot provided by a particle inside a transformation to the world outside a transformation should be explicitly done (via something like `mapSlot`?) and not relying on names alone. It should be possible for particle A (in a transformation) to provide `detail`, which is consumed by B (also in a transformation), while at the same time transformationParticle could provide a `detail` slot, where the name clash is entirely coincidental, and transformationParticle.detail !== A.detail. 

This is why I think we should have some explicit control of this for the transformation particle.

Reply:
BTW are you happy for me to land this as is?
We can continue the conversation while I build on top of this.

Reply:
I was wondering about TransformationParticle that is generic (more like today's Multiplexer) and it doesn't have a hardcoded particle A and B in its code.
For example, if i have:
```
particle ShowProduct
  consume item
    provide annotation

particle Arrivinator
  consume annotation
```
how do I declare TransformationParticle's consume and provide? do i need a separate spec for each of the hosted shapes?
 
and then in the recipe i want to bind the transformed `annotation` slots:
```
recipe
    TransformationParticle
       hostedParticle = ShowProduct
       // consume? provide?
    TransformationParticle
       hostedParticle = Arrivinator
       consume ???
```

Reply:
Interesting. This could be done if we allow a particle to create new slots programmatically outside a transformation. Or if we strategize inside transformations, Arrivinator could be added inside the innerArc.

You can obviously do:
```
recipe
    TransformationParticle
      hosted1 = ShowProduct
      hosted2 = Arrivinator
```
and have TransformationParticle do some processing to connect ShowProduct and Arrivinator.

Reply:
sure, but the slots created by TransformationParticle needs to be available for the next recipes, so we need to have a way to declare them in manifest and reference them in the manifest. it is the manifest format i'm asking about, not the implementation.

Reply:
@piotrswigon TBH this name clash sounds like it's simply a bug in the SlotContext machinery. We don't need explicit control, we just need our systems to work properly.
@mmandlis this is why we need generic type variable holes for names in interfaces; this is something I specced out (at least roughly) a while ago but we haven't need it in practice yet. The other feature that fits in here is being able to specify interfaces with varargs style extra handles.

```
shape TransformationInterface(~a, ~b, ~c)
  in ~a
  consume ~b
    may provide ~c // i.e. this is an optional match

particle TransformationParticle
  host TransformationInterface(~a, ~b, ~c) innerParticle
  in [~a]
  consume ~b
    may provide ~c
```

@piotrswigon let's *not* do this, it introduces a host of potential security leaks that we can't pick up through static data analysis